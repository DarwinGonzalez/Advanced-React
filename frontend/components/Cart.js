import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import SickButton from './styles/SickButton';
import CloseButton from './styles/CloseButton';

const LOCAL_STATE_QUERY = gql`
	query {
		cartOpen @client
	}
`;

const TOGGLE_CART_MUTATION = gql`
	mutation {
		toggleCart @client
	}
`;

const Cart = () => (
	<Mutation mutation={TOGGLE_CART_MUTATION}>
		{toggleCart => (
			<Query query={LOCAL_STATE_QUERY}>
				{({ data }) => (
					<CartStyles open={data.cartOpen}>
						<header>
							<CloseButton title="close" onClick={toggleCart}>
								&times;
							</CloseButton>
							<Supreme>Your Cart</Supreme>
							<p>You Have __ Items in your cart.</p>
						</header>

						<footer>
							<p>$10.10</p>
							<SickButton>Ckeckout</SickButton>
						</footer>
					</CartStyles>
				)}
			</Query>
		)}
	</Mutation>
);

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
