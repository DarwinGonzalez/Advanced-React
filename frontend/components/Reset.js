import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

const RESET_MUTATION = gql`
	mutation RESET_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
	    resetPassword(
            resetToken: $resetToken,
            password: $password,
            confirmPassword: $confirmPassword
        ) {
            id
            email
            name
		}
	}
`;

class Reset extends Component {
    static propTypes = {
        resetToken: PropTypes.string.isRequired
    };
	state = {
        password: '',
        confirmPassword: ''
	};
	saveToState = (e) => {
		this.setState({ [e.target.name]: e.target.value });
	};
	render() {
		return (
            <Mutation
                mutation={RESET_MUTATION}
                refetchQueries={[{query: CURRENT_USER_QUERY }]}
                variables={{
                    resetToken: this.props.resetToken,
                    password: this.props.password,
                    confirmPassword: this.props.confirmPassword
                }}>
				{(reset, { error, loading, called }) => (
					<Form method="post" onSubmit={async (e) => {
                        e.preventDefault();
                        await reset();
                        this.setState({ password: '', confirmPassword: ''});
                    }}>
						<fieldset disabled={loading} aria-busy={loading}>
							<h2>Reset your password</h2>
                            <Error error={error}></Error>
							<label htmlFor="password">
								Password
								<input
									type="password"
									name="password"
									value={this.state.password}
									onChange={this.saveToState}
								/>
							</label>
                            <label htmlFor="confirmPassword">
								Confirm your Password
								<input
									type="password"
									name="confirmPassword"
									value={this.state.confirmPassword}
									onChange={this.saveToState}
								/>
							</label>
							<button type="submit">Request Reset!</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		);
	}
}

export default Reset;
