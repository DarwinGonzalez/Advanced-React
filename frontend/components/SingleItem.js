import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import styled from 'styled-components';
import Error from './ErrorMessage';
import Head from 'next/head';

const SingleItemStyle = styled.div`
    max-width: 1200px;
    margin: 2rem auto;
    box-shadow: ${props => props.theme.bs};
    display: grid;
    grid-auto-columns: 1fr;
    grid-auto-flow: column;
    min-height: 800px;
    img {
        width: 100%;
        height: 100%;
        object-fit: contain;
    }
    .details {
        margin: 3rem;
        font-size: 2rem;
    }
`;

const SINGLE_ITEM_QUERY = gql`
    query SINGLE_ITEM_QUERY($id: ID!) {
        item(where: { id: $id }) {
            id
            title
            description
            largeImage
        }
    }
`;

class SingleItem extends Component {
    render() {
        return (
            <Query query={SINGLE_ITEM_QUERY} variables={{id: this.props.id}}>
                {({error, loading, data}) => {
                    if(error) return <Error error={error} />;
                    if(loading) return <p>Loading...</p>;
                    if(!data.item) return <p>No Item found for {this.props.id} </p>;
                    const item = data.item;
                    return <SingleItemStyle>
                        <Head>
                            <title>Sick fits | { item.title }</title>
                        </Head>
                        <img src={item.largeImage} alt={item.title} />
                        <div className="details">
                            <h2>Viewing {item.title}</h2>
                            <p>{item.descriptioncode}</p>
                        </div>
                    </SingleItemStyle>;
                }}
            </Query>
        )
    }
}

export default SingleItem;
export {SINGLE_ITEM_QUERY};