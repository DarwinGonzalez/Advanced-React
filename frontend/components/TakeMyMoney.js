import React from 'react';
import StripeCheckout from 'react-stripe-checkout';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import Nprogress from 'nprogress';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import calcTotalPrice from '../lib/calcTotalPrice';
import Error from './ErrorMessage';
import User, { CURRENT_USER_QUERY } from './User';

const CREATE_ORDER_MUTATION = gql`
	mutation createOrder($token: String!) {
		createOrder(token: $token) {
			id
			charge
			total
			items {
				id
				title
			}
		}
	}
`;

function totalItems(cart) {
	return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0);
}
class TakeMyMoney extends React.Component {
	onToken = async (res, createOrder) => {
        Nprogress.start();
        // manually call the mutation once we have the stripe token
        const order =  await createOrder({
            variables: {
                token: res.id
            }
        }).catch(err => {
            alert(err.message);
        });
        Router.push({
            pathname: '/order',
            query: { id: order.data.createOrder.id }
        });
	};
	render() {
		return (
			<User>
				{({ data: { me }, loading }) => {
				if(loading) return null;
				return (
					<Mutation mutation={CREATE_ORDER_MUTATION} refetchQueries={[ { query: CURRENT_USER_QUERY } ]}>
						{(createOrder) => (
							<StripeCheckout
								amount={calcTotalPrice(me.cart)}
								name="Sick Fits"
								description={`Order of ${totalItems(me.cart)} item${totalItems(me.cart) === 1
									? ''
									: 's'}!`}
								image={me.cart.length && me.cart[0] && me.cart[0].item.image}
								stripeKey="pk_test_yHqrpIuxxlYQ3TFkzr5jH3be00Nt1J7Xi9"
								currency="USD"
								email={me.email}
								token={(res) => this.onToken(res, createOrder)}
							>
								{this.props.children}
							</StripeCheckout>
						)}
					</Mutation>
				)}}
			</User>
		);
	}
}

export default TakeMyMoney;
export { CREATE_ORDER_MUTATION };