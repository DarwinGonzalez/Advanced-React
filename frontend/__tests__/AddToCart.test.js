import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import AddToCart, { ADD_TO_CART_MUTATION } from '../components/AddToCart';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

const mocks = [
    {
        request: {query: CURRENT_USER_QUERY },
        result: { data: { me: {
            ...fakeUser(),
            cart: []
        }}}
    },
    {
        request: {query: CURRENT_USER_QUERY },
        result: { data: { me: {
            ...fakeUser(),
            cart: [fakeCartItem]
        }}}
    },
    {
        request: { query: ADD_TO_CART_MUTATION, variables: { id: 'abc123 '}},
        result: {
            data: {
                addToCart: {
                    ...fakeCartItem(),
                    quantity: 1,
                }
            }
        }
    }
];

describe('<AddToCart /', () => {
    it('renders and macthes the snapshot', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <AddToCart id="abc123"></AddToCart>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
    });

    it.skip('adds an item to car when clicked', async () => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                {
                    client => {
                        apolloClient = client;
                        return <AddToCart id="abc123"></AddToCart>;
                    }
                }
                </ApolloConsumer>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const { data: { me } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
        expect(me.cart).toHaveLength(0);
        // add an item to the cart
        wrapper.find('button').simulate('click');
        await wait();
        // check if the item is in the cart
        const { data: { me: me2 } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
        expect(me2.cart).toHaveLength(1);
    });

    it('changes from add to adding when clicked', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <AddToCart id="abc123"></AddToCart>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.text()).toContain('Add To Cart');
        wrapper.find('button').simulate('click');
        expect(wrapper.text()).toContain('Adding To Cart');
    });
});