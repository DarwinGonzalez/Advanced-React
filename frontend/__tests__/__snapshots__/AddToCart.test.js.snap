// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart / renders and macthes the snapshot 1`] = `
<button
  disabled={false}
  onClick={[Function]}
>
  Add
   To Cart ðŸ›’
</button>
`;
